src/vmm/src/lib.rs

```diff
@@ -214,17 +219,29 @@ impl VMM {
         let zero_page_addr = GuestAddress(ZEROPG_START);
 
         // Load the kernel into guest memory.
-        let kernel_load = Elf::load(
+        let kernel_load = match Elf::load(
             &self.guest_memory,
             None,
             &mut kernel_image,
             Some(GuestAddress(kernel_cfg.himem_start)),
-        )
-        .map_err(Error::KernelLoad)?;
+        ) {
+            Ok(result) => result,
+            Err(loader::Error::Elf(elf::Error::InvalidElfMagicNumber)) => BzImage::load(
```

Comment:
I think it would be nicer to get the file type and then either parse it as elf or as bzimage. It looks like calling the `file` command gives you the information about it being elf/bzimage, which makes me think that there should be some helpers available in rust as well that we can use.

Reply:
I'm looking into this. Funnily enough, the only discernible way I can find is trial and error. ELF binaries have a magic number in the header (that's what `readelf` checks, `file` too) that `bzImage` files don't - which is obvious enough, as `bzImage` is a kernel-only format, while ELF is pretty much any linked and executable binary, as the name suggests. So, in order to discern an ELF kernel from a `bzImage`, the only thing we can do is:
* read the first 16 bytes and look for the ELF magic number ([source](https://man7.org/linux/man-pages/man5/elf.5.html))
* if it's there, it certainly isn't a `bzImage`, so we can carry on trying to boot it and if it was just an ordinary executable, we'll fail
* if it's missing, follow the kernel format ([source](https://www.kernel.org/doc/Documentation/x86/boot.txt)) and since we know for sure it's not ELF, use the `HdrS` field to discern `bzImage` from `something else`

This is...exactly what we do now, with the difference that we chain the attempts instead of packing them in a separate function.

Reply:
I also took a look at how `file` is able to decide what is the file format when it is printing it. It is using [`magic`](https://linux.die.net/man/5/magic), which looks like reading the header for both elf & bzimage and compare it with a magic header database. This would mean parsing the header twice, and since the parser is already doing that, I suppose we can just leave it as is. Sorry for derailing you.